### PR TITLE
chore: add token contracts

### PR DESCRIPTION
**Pull request by gpsanant**
Adding token contracts to this repo.

There's the gross addition of different versioned submodules.

The difference between whats on mainnet and what's in the repo is solely that the EigenStrategy in the repo includes SafeERC20 from OZ v4.7.1 whereas, on mainnet, it uses v4.9.0. 

This contains no storage and will be up to date after our next upgrade. This is done primarily so the compiler doesn't think there's a ton of redeclared imports.

### CODE REVIEW

File: src/contracts/strategies/EigenStrategy.sol [hunk 1]
```diff
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity =0.8.12;
+
+// NOTE: Mainnet uses the OpenZeppelin v4.9.0 contracts, but this imports the 4.7.1 version. This will be changed after an upgrade.
```
Comment (ChaoticWalrus) [inline src/contracts/strategies/EigenStrategy.sol:4]: so is the plan to upgrade all contracts to 4.9.0 ?
it's definitely ugly to have both versions present at once

**Review by ChaoticWalrus — APPROVED** [review]
Looks OK to me.
I would like to:

- discuss how we will resolve the submodule hell that this introduces
- get the token addresses added to the README table

File: src/contracts/interfaces/IEigen.sol [hunk 2]
```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity >=0.5.0;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
```
Comment (wadealexc) [inline src/contracts/interfaces/IEigen.sol:4]: is this not 4.9.0?

Comment (wadealexc) [reply src/contracts/interfaces/IEigen.sol:4]: not that it should change, but...

File: src/contracts/strategies/EigenStrategy.sol [hunk 3]
```diff
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity =0.8.12;
```
Comment (wadealexc) [inline src/contracts/strategies/EigenStrategy.sol:2]: pragmas need to be `^0.8.12`

File: src/contracts/token/BackingEigen.sol [hunk 4]
```diff
@@ -0,0 +1,132 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity =0.8.12;
```
Comment (wadealexc) [inline src/contracts/token/BackingEigen.sol:2]: pragmas need to be `^0.8.12`

File: src/contracts/token/Eigen.sol [hunk 5]
```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity =0.8.12;
```
Comment (wadealexc) [inline src/contracts/token/Eigen.sol:2]: pragmas need to be `^0.8.12`

File: src/test/harnesses/EigenHarness.sol [hunk 6]
```diff
@@ -0,0 +1,33 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity =0.8.12;
```
Comment (wadealexc) [inline src/test/harnesses/EigenHarness.sol:2]: pragmas need to be `^0.8.12`

File: src/test/token/EigenTransferRestrictions.t.sol [hunk 7]
```diff
@@ -0,0 +1,180 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity =0.8.12;
```
Comment (wadealexc) [inline src/test/token/EigenTransferRestrictions.t.sol:2]: pragmas need to be `^0.8.12`

File: src/test/token/EigenWrapping.t.sol [hunk 8]
```diff
@@ -0,0 +1,200 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity =0.8.12;
```
Comment (wadealexc) [inline src/test/token/EigenWrapping.t.sol:2]: pragmas need to be `^0.8.12`

File: src/test/token/bEIGEN.t.sol [hunk 9]
```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity =0.8.12;
```
Comment (wadealexc) [inline src/test/token/bEIGEN.t.sol:2]: pragmas need to be `^0.8.12`

File: src/contracts/interfaces/IEigen.sol (hunk 2 above)
Comment (wadealexc) [reply src/contracts/interfaces/IEigen.sol:4]: oh you can't do this because `EigenStrategy` is on 4.7.1

ouch

**Review by wadealexc — CHANGES_REQUESTED** [review]
Needs contract addresses in root README

File: script/utils/ExistingDeploymentParser.sol [hunk 10]
```diff
@@ -623,7 +645,7 @@ contract ExistingDeploymentParser is Script, Test {
         for (uint256 i = 0; i < numStrategiesToDeploy; ++i) {
             vm.serializeAddress(deployed_strategies, strategiesToDeploy[i].tokenSymbol, address(deployedStrategyArray[i]));
         }
-        string memory deployed_strategies_output = numStrategiesToDeploy == 0
+        string memory deployed_strategies_output = numStrategiesDeployed == 0
```
Comment (wadealexc) [inline script/utils/ExistingDeploymentParser.sol:648]: is this correct? seems out of place in this PR

File: src/contracts/strategies/EigenStrategy.sol (hunk 1 above)
Comment (gpsanant) [reply src/contracts/strategies/EigenStrategy.sol:4]: We'd need to figure out how to deal with the EIP712 storage change

Comment (wadealexc) [reply src/contracts/strategies/EigenStrategy.sol:4]: i think the issue was primarily when doing a downgrade, upgrades are designed to be possible. but yes, good call - will require some review

File: script/utils/ExistingDeploymentParser.sol (hunk 10 above)
Comment (gpsanant) [reply script/utils/ExistingDeploymentParser.sol]: nope from an intermediate state i got into, should be reversed

File: script/configs/holesky/Holesky_current_deployment.config.json [hunk 11]
```diff
@@ -38,7 +38,16 @@
...
     "strategyManager": "0xdfB5f6CE42aAA7830E94ECFCcAd411beF4d4D5b6",
-    "strategyManagerImplementation": "0x59f766A603C53f3AC8Be43bBe158c1519b193a18"
+    "strategyManagerImplementation": "0x59f766A603C53f3AC8Be43bBe158c1519b193a18",
+    "token": {
+      "tokenProxyAdmin": "0x67482666771e82C9a73BB9e9A22B2B597f448BBf",
+      "EIGEN": "0x3B78576F7D6837500bA3De27A60c7f594934027E",
+      "bEIGEN": "0x275cCf9Be51f4a6C94aBa6114cdf2a4c45B9cb27",
+      "EIGENImpl": "0x083bC9e0DCF2C3e13E24686e5202232995578c5a",
+      "bEIGENImpl": "0x4500927874Ad41538c1bEF2F5278E7a86DF6bce8",
+      "eigenStrategy": "0x43252609bff8a13dFe5e057097f2f45A24387a84",
```
Comment (8sunyuan) [inline script/configs/holesky/Holesky_current_deployment.config.json:48]: I'm seeing on Holesky the EigenStrategy is deployed to `0x43252609bff8a13dFe5e057097f2f45A24387a84` but it has not yet been whitelisted on the StrategyManager `0xdfB5f6CE42aAA7830E94ECFCcAd411beF4d4D5b6`

File: README.md [hunk 12]
```diff
@@ -138,6 +138,8 @@ The current mainnet deployment is our M2 release. You can view the deployed cont
 | [`StrategyBase (mETH)`](https://github.com/Layr-Labs/eigenlayer-contracts/blob/mainnet/src/contracts/strategies/StrategyBaseTVLLimits.sol) | [`0x298aFB19A105D59E74658C4C334Ff360BadE6dd2`](https://etherscan.io/address/0x298aFB19A105D59E74658C4C334Ff360BadE6dd2) | [`0xdfdA...46d3`](https://etherscan.io/address/0xdfdA04f980bE6A64E3607c95Ca26012Ab9aA46d3) | Proxy: [`TUP@4.7.1`](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.7.1/contracts/proxy/transparent/TransparentUpgradeableProxy.sol) |
 | `Beacon Chain ETH` | `0xbeaC0eeEeeeeEEeEeEEEEeeEEeEeeeEeeEEBEaC0` | - | - Used for Beacon Chain ETH shares <br />- Not a real contract! |
 
+###### Strategies - EIGEN
+
```
Comment (wadealexc) [inline README.md:142]: This table creation did not work. you need a header

**Review by wadealexc — CHANGES_REQUESTED** [review]
deploy addrs need a section for EIGEN/bEIGEN token addrs

**Review by wadealexc — CHANGES_REQUESTED** [review]
Implementation addrs in deploy tables are now truncated (see every other address in that table). We're doing 2 bytes on each side: `0x1122...3344`

File: script/configs/holesky/Holesky_current_deployment.config.json (hunk 11 above)
Comment (gpsanant) [reply script/configs/holesky/Holesky_current_deployment.config.json:48]: should be whitelisted now https://holesky.etherscan.io/tx/0xfed1f9687aa3b1300dfb3142986e7c0dd06fdf9b360657bf8d9c022efdf38b15

**Review by wadealexc — APPROVED** [review]

**Pull request merged**